### ISSUE TITLE
Azure DevOps support

### ISSUE THREAD

**Issue by lafe** [Azure/static-web-apps#5]
Static Web Apps is something that we would be able to use in our customer projects. However, I am confused regarding the support for Azure DevOps. Currently, there seems to be a focus on GitHub but there is no mention of Azure DevOps. Is support for Azure DevOps planned in the future?

**Comment by zunderscore** [comment]
Wondering the same thing. Love GitHub, but I do all of my work, store all my repos, setup my CI/CD, etc. in Azure DevOps.

**Comment by Swimburger** [comment]
I'm wondering the same, is there support to deploy Static Web Apps from command line and any CI/CD system in the future?

**Comment by eparizzi** [comment]
I'm just wondering why isn't a basic CLI is first built? Then, it can be re-used in any service, GitHub Actions, Azure DevOps pipelines or even anything that can run a command line.

**Comment by fudgeboy2000** [comment]
Same here - would be really exciting to host Blazor WASM apps in Static Sites, but all our work is done via Azure DevOps.  Why restrict this to GitHub?

**Comment by radumg** [comment]
+1, please add details regarding Azure DevOps support as we already heavily use AppService with static web apps but all our CI/CD and repos are in DevOps.

**Comment by Rutix** [comment]
https://twitter.com/AbelSquidHead/status/1263265901509439489 seems to suggest that Azure Pipelines are supported but I still wonder if Azure Repos are going to be supported or not.

**Comment by mahmoudajawad** [comment]
How can it be supported when at time of creating a Static Web App it requires Github auth? Has anyone found a method to create SWA resource from CLI that can be linked to Repos or Pipelines?

**Comment by Rutix** [comment]
I think because you can use Azure Pipelines with a Github repo that it is supported in that way. I don't think it supports Azure Repos for now (it's still in preview, after all, there are all kinds of constraints).

**Comment by Tiberriver256** [comment]
I'd heard rumors that Azure DevOps was losing focus and Github was going to be the focus moving forward... this kind of confirms the direction for me. Wish it would be made public though.

**Comment by mahmoudajawad** [comment]
@Tiberriver256 Microsoft is well known for being committed to enterprises and governments. Enterprises currently using ADO would not like moving to primitive-all-issues from the great current tools of Azure Boards and how all things get linked together with Azure Pipelines and Azure Repos. For that, I highly doubt this is going to happen, unless Microsoft moves all ADO features to Github, which again doesn't make sense, since corporates usually tend to merge new acquisitions into the parent branding, not empowering the acquired brands. However, this is all a personal opinion and I might be completely wrong.

**Comment by anganti** [comment]
For preview, GitHub Actions is the only supported way to build and deploy to StaticWebApps. 

Supporting other CI/CD pipelines and/or CLIs is currently work in progress. 

Thank you for your feedback. Keep them coming.

**Comment by laurentbel** [comment]
Definitively needed !
Idea submitted here as well : https://feedback.azure.com/forums/169385-web-apps/suggestions/40521196-static-web-apps-support-for-generic-scm-and-ci-cd

**Comment by sushruth** [comment]
@anganti Is there any place we could track its progress?

**Comment by brentos99** [comment]
Was super excited for Azure Static websites as it is exactly what we need at the moment. But we are committed to azure dev ops.

**Comment by mahmoudajawad** [comment]
This might not be the best approach for all, but for me it is working fine--I've now having my pipeline of building the artifacts and push them as static files to a GH repo which is then used to deploy SWA.

**Comment by chrisd08** [comment]
Is there a way to make the Github build action at least run tests before deploying to azure?

**Comment by lucaslra** [comment]
+1 on this.
This should be there from the start, to be honest.
Having to resort to Storage Static Sites for now, but definitely not an ideal solution.

**Comment by litan1106** [comment]
+1 on this.

**Comment by pherbel** [comment]
+1

**Comment by erosnemesis** [comment]
+1 - After watching some of the .NET Conf today I was looking to deploy my simple app to a Static Web Site. I mainly use Azure DevOps (especially at work) and am hoping to get my company on-board with Blazor WASM, but I need a full proof of concept with pipelines. This is a huge must and surprised that GitHub came first.

**Comment by morvaizo** [comment]
+1

**Comment by FrankBurmo** [comment]
@annaji-msft any update on the progress to implement devops-support?

**Comment by anthonychu** [comment]
Maybe in a couple of months. Some of the platform work and Pipeline task are getting close but will take some time to get the changes into the portal. We may be able to share some rough instructions using the CLI when that is ready.

**Comment by daKmoR** [comment]
it's a new year - are there any updates? 🤔

Maybe an API endpoint where we can push a zip file would be good enough... we can add it to the azure pipeline ourselves 😬
or a CLI we can execute in the pipeline

**Comment by anthonychu** [comment]
The portal work items have been completed and should be deployed over the next month or so. The Azure DevOps task is also complete and will need some time to be added to the list of tasks.

Given what we know right now, the ETA for Azure DevOps support is February. We’ll update here with docs when ready.

Zip deployment should be possible in the coming months as well.

**Comment by erikkroes** [comment]
Thanks for the update. We're really looking forward to this! I've heard static is this amazing new trend 👀

**Comment by IhorLeontiev** [comment]
Thanks for the information. Waiting a lot for this integration. This is the only thing that stops us from using Static Web App instead of Azure Static WebSite on Storage

**Comment by anthonychu** [comment]
Update on where we are today...

We're waiting for our Azure DevOps task to be published. There's still no solid ETA on this yet but hoping within the next week or so.

The portal work is ready. It'll take a week or two to make it live after the task is published.

And to set expectations, what you'll be able to do (hopefully later this month) is:
* Create a Static Web App without linking it to GitHub
* Add an Azure Pipelines YAML to your repo with the task (we'll provide an example YAML that you can copy/paste). The task looks a lot like the GitHub Action in terms of input and capabilities. You'll need to set a pipeline variable for the deployment token (retrieved from the portal).
* No support for PRs/environments yet. We don't have the same hooks as GitHub to act on PR open/close. We'll continue to investigate.

We'll also work on improving the experience in the coming months. But we're trying to get the basic functionality out first so most of you are unblocked ASAP. Thanks for your patience on this.

**Comment by blueboxes** [comment]
Will this also bring out closer to being out of preview?

**Comment by litan1106** [comment]
@anthonychu awesome, thanks. I like to work with Azure Pipeline not Github Actions.

**Comment by daKmoR** [comment]
@anthonychu really love those updates 🤗 

would it be possible to get an update roughly every 7 or 14 days? 🤔  it seems like a highly anticipated feature and I guess we all would appreciate it - thxxx 🙇‍♂️

**Comment by anthonychu** [comment]
Unfortunately, I don't really have much to update at this time. We're still waiting for an ETA on when the task will be published.

**Comment by vmehmeri** [comment]
@anthonychu  Sorry for insisting, but no ETA yet or any update?

**Comment by devgibsonsp** [comment]
Really hoping for this to be available soon...

**Comment by anthonychu** [comment]
Task is rolling out! If all goes well, UI changes to support this would be live mid-March.

**Comment by DavidNorena** [comment]
@anthonychu you and your team mates are amazing for real, we've been waiting for almost a year for this, its so nice that finally we'll see it shining, AMAZING JOB TEAM !

**Comment by vmanickamus** [comment]
Hi do we have a ETA for this feature release? Thanks.

**Comment by loganmarshall1** [comment]
> @anthonychu you and your team mates are amazing for real, we've been waiting for almost a year for this, its so nice that finally we'll see it shining, AMAZING JOB TEAM !

Agreed! Thank you so much @anthonychu ! Looking forward to this! 👍🏼

**Comment by MichelZ** [comment]
https://github.com/microsoft/azure-pipelines-tasks/tree/master/Tasks/AzureStaticWebAppV0

Here's the task if anyone's interested

**Comment by hrazmsft** [comment]
Can we use it now without GitHub repo and it will work or we have to wait for the new UI?

**Comment by elahnala** [comment]
This is totally work in progress. Developer experience is quite weak with this solution. Seems like the tools have zero documentation and totally in beta-phase or even alpha.

**Comment by MichelZ** [comment]
I haven't tested it yet, so I don't know...  But it can't be much longer now until it's "officially" supported :)

**Comment by MichelZ** [comment]
> 
> 
> This is totally work in progress. Developer experience is quite weak with this solution. Seems like the tools have zero documentation and totally in beta-phase or even alpha.

You've discovered what the label "Preview" means...  congratulations! ;)

**Comment by Rutix** [comment]
@vmehmeri In the post above by @anthonychu it says this:

"You'll need to set a pipeline variable for the deployment token (retrieved from the portal)." 

So I figure a change to the "Static Web App" UI in the azure portal isn't live yet.

**Comment by TroyWitthoeft** [comment]
I've been waiting for this too!  Going to be trying it out today!  My org is all in on Azure DevOps.  Let's see if we can make Azure Static Web Apps play nice with DevOps repos.

**Comment by anthonychu** [comment]
Hey folks. Yes, as you've discovered, the task is rolling out to Azure DevOps. It's a bit of a chicken or egg thing to coordinate portal changes and docs with a rolling release.

I'll update this thread once the task is completely rolled out and we've validated that things are working. Goal is to provide instructions on how to enable this without the portal experience, so those of you on here who are super eager to get started can do so ASAP. We'll see if I can provide those this week. 😅

We'll update the portal UI to show the new experience once we're done with validation. Look for it to appear near the end of this month.

Official docs will roll out around the same time as the portal changes. Though expect the same situation where it may appear in one place before the other.

**Comment by TroyWitthoeft** [comment]
@anthonychu Our Azure portal tenant still doesn't have the new Azure Static Web Apps UI enabled.  Waiting on the feature flag. 

You mentioned at the [31m mark](https://www.youtube.com/watch?v=xEYQBPo11iI&t=31m0s) of the video that there may be a workaround?  We REALLY REALLY want to use it.  😀   Mind sharing?

**Comment by Rutix** [comment]
> @anthonychu Our Azure portal tenant still doesn't have the new Azure Static Web Apps UI enabled. Waiting on the feature flag.
> 
> You mentioned at the [31m mark](https://www.youtube.com/watch?v=xEYQBPo11iI&t=31m0s) of the video that there may be a workaround? We REALLY REALLY want to use it. 😀 Mind sharing?

The video also mentions that they will update the docs with cli commands to do the same. Let's just be patience and let @anthonychu have the time to update the docs 😄.

**Comment by anthonychu** [comment]
Okay, here are some rough CLI instructions. Thanks for patiently waiting. Keep in mind you'll likely see some inconsistencies in the portal when you do this. This is normal as the changes to support some of this stuff are not rolled out yet.

```
az group create -n test-azuredevops -l westus2
az staticwebapp create -l westus2 -n test-azuredevops -g test-azuredevops -s "" -b "" --token "-"
```

We'll update the CLI to make those flags optional. For now, use the above values to bypass validation.

Go to your Azure Repo and create a pipeline:

![image](https://user-images.githubusercontent.com/3982077/111025349-681e0e80-8398-11eb-9ad0-1d5cfc1e7187.png)

![image](https://user-images.githubusercontent.com/3982077/111025359-753afd80-8398-11eb-9526-5f63b9e8dd77.png)

Update the workflow to add the Static Web Apps task, entering the correct values for your app, api, and output locations.

```
trigger:
  - main
pool:
  vmImage: ubuntu-latest
steps:
  - task: AzureStaticWebApp@0
    inputs:
      app_location: frontend 
      api_location: api
      output_location: build
    env:
      azure_static_web_apps_api_token: $(deployment_token)
```

Grab the deployment token from the Azure portal and add a secret variable named `deployment_token`.

![image](https://user-images.githubusercontent.com/3982077/111025438-fa261700-8398-11eb-9963-80a152004157.png)

---

![image](https://user-images.githubusercontent.com/3982077/111025441-0316e880-8399-11eb-922d-f427210d4721.png)

Save and run!

**Comment by exhaler** [comment]
> Okay, here are some rough CLI instructions. Thanks for patiently waiting. Keep in mind you'll likely see some inconsistencies in the portal when you do this. This is normal as the changes to support some of this stuff are not rolled out yet.
> 
> **I should explicitly mention that this is not the final experience. I'm sharing these instructions for folks who really want to try this (@TroyWitthoeft!) and can't wait for the portal experience to be rolled out** 😅
> 
> ```
> az group create -n test-azuredevops -l westus2
> az staticwebapp create -l westus2 -n test-azuredevops -g test-azuredevops -s "" -b "" --token "-"
> ```
> 
> We'll update the CLI to make those flags optional. For now, use the above values to bypass validation.
> 
> Go to your Azure Repo and create a pipeline:
> 
> ![image](https://user-images.githubusercontent.com/3982077/111025349-681e0e80-8398-11eb-9ad0-1d5cfc1e7187.png)
> 
> ![image](https://user-images.githubusercontent.com/3982077/111025359-753afd80-8398-11eb-9526-5f63b9e8dd77.png)
> 
> Update the workflow to add the Static Web Apps task, entering the correct values for your app, api, and output locations.
> 
> ```
> trigger:
>   - main
> pool:
>   vmImage: ubuntu-latest
> steps:
>   - task: AzureStaticWebApp@0
>     inputs:
>       app_location: frontend 
>       api_location: api
>       output_location: build
>     env:
>       azure_static_web_apps_api_token: $(deployment_token)
> ```
> 
> Grab the deployment token from the Azure portal and add a secret variable named `deployment_token`.
> 
> ![image](https://user-images.githubusercontent.com/3982077/111025438-fa261700-8398-11eb-9963-80a152004157.png)
> 
> ![image](https://user-images.githubusercontent.com/3982077/111025441-0316e880-8399-11eb-922d-f427210d4721.png)
> 
> Save and run!

Finally! However, when I run ```az staticwebapp create -l westus2 -n test-azuredevops -g test-azuredevops -s "" -b "" --token "-"``` i get an error ```argument --source/-s: expected one argument```

```az --version
azure-cli                         2.20.0

core                              2.20.0
telemetry                          1.0.6

Extensions:
interactive                        0.4.4

Python location 'C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\python.exe'
Extensions directory 'C:\Users\Exhaler\.azure\cliextensions'

Python (Windows) 3.6.8 (tags/v3.6.8:3c6b436a57, Dec 23 2018, 23:31:17) [MSC v.1916 32 bit (Intel)]

Legal docs and information: aka.ms/AzureCliLegal


Your CLI is up-to-date.

Please let us know how we are doing: https://aka.ms/azureclihats```

**Comment by anthonychu** [comment]
@exhaler Interesting. Might be a problem on Windows. Try in https://shell.azure.com/ (Just tested and worked for me)

**Comment by smurawski** [comment]
@exhaler @anthonychu  Are you running that from PowerShell?  I experienced similar behavior and I think it's due to the double pass of string evaluation - PowerShell evaluating the string (effectively dropping the quotes as it passes it to the CLI) then the cli's runtime tries to evaluate it.

There are two options - use the stop-parsing symbol to tell PowerShell to ignore evaluation past a certain point. [about_Parsing](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_parsing?view=powershell-7.1)

`az staticwebapp create --% -l westus2 -n test-azuredevops -g test-azuredevops -s "" -b "" --token "-"`

Or a much more involved approach with custom PowerShell wrappers (which we did in Chef - https://github.com/chef/chef/pull/3339 and is probably overkill, but made for the customer experience we wanted) .

**Comment by exhaler** [comment]
> @exhaler @anthonychu Are you running that from PowerShell? I experienced similar behavior and I think it's due to the double pass of string evaluation - PowerShell evaluating the string (effectively dropping the quotes as it passes it to the CLI) then the cli's runtime tries to evaluate it.

I tried both PowerShell and Command prompt (cmd) which gave me the same error result. Everything worked correctly when I used https://shell.azure.com/

**Comment by IhorLeontiev** [comment]
@exhaler  @smurawski everything worked well on Bash via https://shell.azure.com/, but in Powershell:
the following arguments are required: --branch/-b

BTW @anthonychu is it possible to have an ARM example?

**Comment by smurawski** [comment]
@IhorLeontiev In PowerShell, you have to either use the stop parsing symbol `--%` before the parameters or put a second pair of quotes around the branch and source parameters, eg. single quotes wrapping the double quotes `'""'`.

For ARM you can deploy something like
```
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "name": {
            "type": "string"
        },
        "location": {
            "type": "string"
        }
    },
    "resources": [
        {
            "apiVersion": "2019-12-01-preview",
            "name": "[parameters('name')]",
            "type": "Microsoft.Web/staticSites",
            "location": "[parameters('location')]",
            "tags": null,
            "properties": {},
            "sku": {
                "Tier": "Free",
                "Name": "Free"
            }
        }
    ],
    "outputs": {
        "static_site": {
            "type": "object",
            "value": "[reference(resourceId('Microsoft.Web/staticSites',parameters('name')))]"
        }
    }
}
```

**Comment by vikaskk13** [comment]
ngcc --properties es2015 browser module main --first-only

Error: EACCES: permission denied, open '/working_dir/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__'
    at Object.openSync (node:fs:488:3)
    at Object.writeFileSync (node:fs:1512:35)
    at NodeJSFileSystem.writeFile (/working_dir/node_modules/@angular/compiler-cli/src/ngtsc/file_system/src/node_js_file_system.js:131:16)
    at LockFileWithChildProcess.write (/working_dir/node_modules/@angular/compiler-cli/ngcc/src/locking/lock_file_with_child_process/index.js:58:21)
    at AsyncLocker.<anonymous> (/working_dir/node_modules/@angular/compiler-cli/ngcc/src/locking/async_locker.js:82:65)
    at step (/working_dir/node_modules/tslib/tslib.js:143:27)
    at Object.next (/working_dir/node_modules/tslib/tslib.js:124:57)
    at /working_dir/node_modules/tslib/tslib.js:117:75
    at new Promise (<anonymous>)
    at Object.__awaiter (/working_dir/node_modules/tslib/tslib.js:113:16)

**Comment by anthonychu** [comment]
@vikaskk13 Looks like a problem specific to Angular. Can you please create a new issue?

**Comment by TeguSolutions** [comment]
> Okay, here are some rough CLI instructions. Thanks for patiently waiting. Keep in mind you'll likely see some inconsistencies in the portal when you do this. This is normal as the changes to support some of this stuff are not rolled out yet.
> 
> **I should explicitly mention that this is not the final experience. I'm sharing these instructions for folks who really want to try this (@TroyWitthoeft!) and can't wait for the portal experience to be rolled out** 😅
> 
> ```
> az group create -n test-azuredevops -l westus2
> az staticwebapp create -l westus2 -n test-azuredevops -g test-azuredevops -s "" -b "" --token "-"
> ```
> 
> We'll update the CLI to make those flags optional. For now, use the above values to bypass validation.
> 
> Go to your Azure Repo and create a pipeline:
> 
> ![image](https://user-images.githubusercontent.com/3982077/111025349-681e0e80-8398-11eb-9ad0-1d5cfc1e7187.png)
> 
> ![image](https://user-images.githubusercontent.com/3982077/111025359-753afd80-8398-11eb-9526-5f63b9e8dd77.png)
> 
> Update the workflow to add the Static Web Apps task, entering the correct values for your app, api, and output locations.
> 
> ```
> trigger:
>   - main
> pool:
>   vmImage: ubuntu-latest
> steps:
>   - task: AzureStaticWebApp@0
>     inputs:
>       app_location: frontend 
>       api_location: api
>       output_location: build
>     env:
>       azure_static_web_apps_api_token: $(deployment_token)
> ```
> 
> Grab the deployment token from the Azure portal and add a secret variable named `deployment_token`.
> 
> ![image](https://user-images.githubusercontent.com/3982077/111025438-fa261700-8398-11eb-9963-80a152004157.png)
> 
> ![image](https://user-images.githubusercontent.com/3982077/111025441-0316e880-8399-11eb-922d-f427210d4721.png)
> 
> Save and run!

I would like to ask is it supposed to be working?
It is the same as the current ms documentation: https://docs.microsoft.com/en-us/azure/static-web-apps/publish-devops
But the following error comes up:
---End of Oryx build logs---
The app build failed to produce artifact folder: '/WebApp/bin/Release/net5.0/wwwroot'. Please ensure this property is configured correctly in your workflow file.

**Comment by techaimail** [comment]
Deploy on CI/CD (in release pipeline) has an incomplete task for deploy called "Static Web App" but there is not field for the token in order to deploy the contents in the right place. When will this overcome the alpha-beta undefined status? static web sites is a really important feature today.

**Comment by anthonychu** [comment]
@techaimail Can you clarify the problem? Is it because the deployment token is an environment variable instead of an input parameter to the task?

**Comment by anthonychu** [comment]
In case you all didn’t see the announcement, the portal experience is live and you can create a Static Web App without linking to GitHub. Documentation is [here](https://docs.microsoft.com/en-us/azure/static-web-apps/publish-devops).

**Comment by smholvoet** [comment]
Gave it a quick try yesterday and this really is a breeze to setup 👍, amazing for stuff like hosting your docs as code somewhere.

**Comment by Basaingeal** [comment]
Does the Azure DevOps task only work on Linux? I'm trying to run it on a Windows self-hosted build server, and it's failing without any information. When I enable system diagnostics, I get this. Seems as though it fails when it can't find bash on the build server.

```
2021-04-01T15:56:03.5147864Z ##[debug]Evaluating condition for step: 'Copy app to Static Web Site'
2021-04-01T15:56:03.5149726Z ##[debug]Evaluating: SucceededNode()
2021-04-01T15:56:03.5150161Z ##[debug]Evaluating SucceededNode:
2021-04-01T15:56:03.5150781Z ##[debug]=> True
2021-04-01T15:56:03.5151262Z ##[debug]Result: True
2021-04-01T15:56:03.5151840Z ##[section]Starting: Copy app to Static Web Site
2021-04-01T15:56:03.5296574Z ==============================================================================
2021-04-01T15:56:03.5296974Z Task         : Deploy Azure Static Web App
2021-04-01T15:56:03.5297290Z Description  : [PREVIEW] Build and deploy an Azure Static Web App
2021-04-01T15:56:03.5297587Z Version      : 0.1.0
2021-04-01T15:56:03.5297811Z Author       : Microsoft Corporation
2021-04-01T15:56:03.5298102Z Help         : https://aka.ms/swadocs
2021-04-01T15:56:03.5298420Z ==============================================================================
2021-04-01T15:56:03.5470874Z ##[debug]Using node path: C:\agent\externals\node\bin\node.exe
2021-04-01T15:56:04.1945072Z ##[debug]agent.TempDirectory=C:\agent\_work\_temp
2021-04-01T15:56:04.1947231Z ##[debug]loading inputs and endpoints
2021-04-01T15:56:04.1948575Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_SYSTEMVSSCONNECTION_ACCESSTOKEN
2021-04-01T15:56:04.1949931Z ##[debug]loading ENDPOINT_AUTH_SCHEME_SYSTEMVSSCONNECTION
2021-04-01T15:56:04.1951312Z ##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
2021-04-01T15:56:04.1952470Z ##[debug]loading INPUT_APP_LOCATION
2021-04-01T15:56:04.1953500Z ##[debug]loading SECRET_SWA-DEPLOYMENT-TOKEN
2021-04-01T15:56:04.1954729Z ##[debug]loading SECRET_SYSTEM_ACCESSTOKEN
2021-04-01T15:56:04.1955856Z ##[debug]loaded 6
2021-04-01T15:56:04.1956749Z ##[debug]Agent.ProxyUrl=undefined
2021-04-01T15:56:04.1957651Z ##[debug]Agent.CAInfo=undefined
2021-04-01T15:56:04.1958530Z ##[debug]Agent.ClientCert=undefined
2021-04-01T15:56:04.1959477Z ##[debug]Agent.SkipCertValidation=undefined
2021-04-01T15:56:04.1960503Z ##[debug]check path : C:\agent\_work\_tasks\AzureStaticWebApp_18aad896-e191-4720-88d6-8ced4806941a\0.1.0\task.json
2021-04-01T15:56:04.1961661Z ##[debug]adding resource file: C:\agent\_work\_tasks\AzureStaticWebApp_18aad896-e191-4720-88d6-8ced4806941a\0.1.0\task.json
2021-04-01T15:56:04.1962710Z ##[debug]system.culture=en-US
2021-04-01T15:56:04.1963561Z ##[debug]which 'bash'
2021-04-01T15:56:04.1964395Z ##[debug]not found
2021-04-01T15:56:04.1965373Z ##[debug]task result: Failed
2021-04-01T15:56:04.1967833Z ##[debug]Processed: ##vso[task.complete result=Failed;]
2021-04-01T15:56:04.1972495Z ##[section]Finishing: Copy app to Static Web Site
```

**Comment by anthonychu** [comment]
@Basaingeal  Yes it currently has a dependency on bash and Docker. We'll update the docs to mention this.

If we're able to remove the bash dependency, do you have Docker installed on your agent?

/cc @miwebst

**Comment by gteirlinck-cigp** [comment]
I also tried to use the DevOps release pipeline task, but I can't seem to find a way to pass the deployment token env variable to the Docker run, which causes the task to fail:

```
2021-04-02T03:36:27.5865921Z Status: Downloaded newer image for mcr.microsoft.com/appsvc/staticappsclient:stable
2021-04-02T03:36:34.9328556Z [37mDeploymentId: be651ec8-08ae-46ac-b00d-77dbb6723ddb[0m
2021-04-02T03:36:34.9336713Z [37m[0m
2021-04-02T03:36:35.0121319Z [31mazure_static_web_apps_api_token was not provided.[0m
2021-04-02T03:36:35.0122436Z [37mThe azure_static_web_apps_api_token is required for deploying content. If you'd like to continue the workflow run without deployment, add the workflow configuration skip_deploy_on_missing_secrets set to true[0m
2021-04-02T03:36:35.0127291Z [37m[0m
2021-04-02T03:36:35.0128129Z [33mFor further information, please visit the Azure Static Web Apps documentation at https://docs.microsoft.com/en-us/azure/static-web-apps/[0m
```

Looking at the task definition, it expects the `azure_static_web_apps_api_token` env variable as input:
```
docker run \
    -e INPUT_AZURE_STATIC_WEB_APPS_API_TOKEN="$azure_static_web_apps_api_token" \
    -e GITHUB_WORKSPACE=$workspace \
[...]
```

However this variable is not mentioned in the `task.json` or in the `index.ts`:

```
const appLocation: string = tl.getInput('app_location', false) || "";
const appBuildCommand: string = tl.getInput('app_build_command', false) || "";
const outputLoction: string = tl.getInput('output_location', false) || "";
const apiLocation: string = tl.getInput('api_location', false) || "";
const apiBuildCommand: string = tl.getInput('api_build_command', false) || "";
const routesLocation: string = tl.getInput('routes_location', false) || "";

process.env['SWA_APP_LOCATION'] = appLocation;
process.env['SWA_APP_BUILD_COMMAND'] = appBuildCommand;
process.env['SWA_OUTPUT_LOCATION'] = outputLoction;
process.env['SWA_API_LOCATION'] = apiLocation;
process.env['SWA_API_BUILD_COMMAND'] = apiBuildCommand;
process.env['SWA_ROUTES_LOCATION'] = routesLocation;
process.env['SWA_DEPLOYMENT_CLIENT'] = deploymentClient;
```

Should it be added there? Otherwise, how can we pass the deploy token value?

**Comment by anthonychu** [comment]
Please see the [docs](https://docs.microsoft.com/en-us/azure/static-web-apps/publish-devops) for how to use the task. We’ll look into adding the token as an input as well.

**Comment by Basaingeal** [comment]
@anthonychu Yes, Docker is installed on the build servers, so if the bash dependency were removed, I think we'd be able to use the task.

**Comment by jballe** [comment]
Thank you for this enhancement!

Have anyone been able to use this in a deployment stage?
I would like to be able to push the same artifact to different webapps/environments, so I have the steps listed below. The task is starting correctly but I get this error message (seems like docker image is correctly pulled and started):

> App Directory Location: '/web' is invalid. Could not detect this directory. Please verify your workflow reflects your repository structure.

My pipeline step is the following. I have also tried using ``app_location: web`` (without the slash prefix) with same result

```
    - task: DownloadPipelineArtifact@2
      displayName: Download web artifact 
      inputs:
        artifact: web
        path: $(Pipeline.Workspace)/web

    - task: DownloadPipelineArtifact@2
      displayName: Download cms artifact 
      inputs:
        artifact: cms
        path: $(Pipeline.Workspace)/web/admin
        patterns: '**'

    - task: DownloadPipelineArtifact@2
      displayName: Download cfg artifact 
      inputs:
        artifact: cfg
        path: $(Pipeline.Workspace)/web
        patterns: '*.json'

    - task: AzureStaticWebApp@0
      inputs:
        app_location: /web
        output_location: ""
      env:
        azure_static_web_apps_api_token: $(webapp_deployment_token)
        BUILD_SOURCESDIRECTORY: $(Pipeline.Workspace)
```

Any hints?
Thank you!

**Comment by anthonychu** [comment]
@jballe Could you please create a new issue so we can track this separately?

**Comment by jballe** [comment]
@anthonychu, sure I just thought it was so closely related to the preview that it would make sense here.

I have tested it a bit more and concluded that there is a ``$(Build.SourcesDirectory)`` also during a deplotment job, so I just had to download the artifacts to a structure within that folder.
If interested you can [take a look at our deployment stage](https://dev.azure.com/kragefolket/_git/website?path=%2Fbld%2Fpipelines%2Ftemplates%2Fdeploy.yml)

Thank you very much for this Azure Pipelines support.

**Comment by tylerewillis** [comment]
Is there any way to set up deployment from ADO in pipeline releases, not just the main pipeline builds?

**Comment by anthonychu** [comment]
@tylerewillis That doesn’t work right now. Can you please file a separate issue to track this?

**Comment by josendf-basetis** [comment]
Hi @tylerewillis,

As a workaround, we have managed to deploy from a DevOps Release Pipeline using a Bash Script task with the following script:

```
workspace="/working_dir"

SWA_OUTPUT_LOCATION=dist
SWA_APP_LOCATION=dist
SWA_DEPLOYMENT_CLIENT=mcr.microsoft.com/appsvc/staticappsclient:stable

docker run \
    -e INPUT_AZURE_STATIC_WEB_APPS_API_TOKEN="$(DEPLOYMENT_TOKEN)" \
    -e GITHUB_WORKSPACE=$workspace \
    -e DEPLOYMENT_PROVIDER=DevOps \
    -e REPOSITORY_URL="$BUILD_REPOSITORY_URI" \
    -e IS_PULL_REQUEST=false \
    -e BASE_BRANCH="$BUILD_SOURCEBRANCHNAME" \
    -e INPUT_APP_LOCATION="$SWA_APP_LOCATION" \
    -e INPUT_OUTPUT_LOCATION="$SWA_OUTPUT_LOCATION" \
    -v "$(System.DefaultWorkingDirectory):$workspace" \
    "$SWA_DEPLOYMENT_CLIENT" \
    ./bin/staticsites/StaticSitesClient upload

```

Also you have to define the release pipeline variable `DEPLOYMENT_TOKEN` with the value of the Static Web App deployment token.

**Comment by exhaler** [comment]
> Hi @tylerewillis,
> 
> As a workaround, we have managed to deploy from a DevOps Release Pipeline using a Bash Script task with the following script:
> 
> ```
> workspace="/working_dir"
> 
> SWA_OUTPUT_LOCATION=dist
> SWA_APP_LOCATION=dist
> SWA_DEPLOYMENT_CLIENT=mcr.microsoft.com/appsvc/staticappsclient:stable
> 
> docker run \
>     -e INPUT_AZURE_STATIC_WEB_APPS_API_TOKEN="$(DEPLOYMENT_TOKEN)" \
>     -e GITHUB_WORKSPACE=$workspace \
>     -e DEPLOYMENT_PROVIDER=DevOps \
>     -e REPOSITORY_URL="$BUILD_REPOSITORY_URI" \
>     -e IS_PULL_REQUEST=false \
>     -e BASE_BRANCH="$BUILD_SOURCEBRANCHNAME" \
>     -e INPUT_APP_LOCATION="$SWA_APP_LOCATION" \
>     -e INPUT_OUTPUT_LOCATION="$SWA_OUTPUT_LOCATION" \
>     -v "$(System.DefaultWorkingDirectory):$workspace" \
>     "$SWA_DEPLOYMENT_CLIENT" \
>     ./bin/staticsites/StaticSitesClient upload
> ```
> 
> Also you have to define the release pipeline variable `DEPLOYMENT_TOKEN` with the value of the Static Web App deployment token.

What would I need to change if my directory looks like this: www/dist/<app files>

**Comment by RossMurr4y** [comment]
You can also pass the parameters for StaticSitesClient directly to the binary inside of the container. 

Environment variables are nice and clean for readability in a pipeline, but I had need to do a manual deployment from a local build I'd done and found the following to be a bit faster to put together.

**Review valid parameters**
```terminal
docker run --entrypoint "/bin/staticsites/StaticSitesClient" \
  mcr.microsoft.com/appsvc/staticappsclient:stable upload --help
```

**Example of a manual deployment of a local build**
```terminal
docker run --entrypoint "/bin/staticsites/StaticSitesClient" \
  --volume "$(pwd)"/build:/root/build \
  mcr.microsoft.com/appsvc/staticappsclient:stable \
  upload \
  --skipAppBuild true \
  --app /root/build \
  --apiToken <token>
```

**Comment by twoCansOfTomatoes** [comment]
Now with GA, can we expect the Azure Static WebApp be decoupled from github and give provision to have other source repos too ?  Also, an option to build, package and deploy separately without having to have the source control configured in the service.

**Comment by anthonychu** [comment]
@twoCansOfTomatoes I think what you're asking for are tracked at #298 (other CI/CD providers) and #7 (deploy from CLI). Can you see if they are what you're looking for and add some comments about what git/CI/CD providers you're thinking of and how you want to deploy without configuring source control? If neither of those apply, please create a new issue with more details. Thanks.

**Comment by miwebst** [comment]
> You can also pass the parameters for StaticSitesClient directly to the binary inside of the container.
> 
> Environment variables are nice and clean for readability in a pipeline, but I had need to do a manual deployment from a local build I'd done and found the following to be a bit faster to put together.
> 
> **Review valid parameters**
> 
> ```
> docker run --entrypoint "/bin/staticsites/StaticSitesClient" \
>   mcr.microsoft.com/appsvc/staticappsclient:stable upload --help
> ```
> 
> **Example of a manual deployment of a local build**
> 
> ```
> docker run --entrypoint "/bin/staticsites/StaticSitesClient" \
>   --volume "$(pwd)"/build:/root/build \
>   mcr.microsoft.com/appsvc/staticappsclient:stable \
>   upload \
>   --skipAppBuild true \
>   --app /root/build \
>   --apiToken <token>
> ```

@RossMurr4y be careful of running the underlying container directly. We will not make breaking changes to our GitHub Action or DevOps task, but we may make breaking changes to the underlying container.

**Comment by dondracek** [comment]
I'm trying to deploy my built app to static webapps using the release pipeline, but I keep getting the error

```
2021-06-05T14:43:43.8414807Z docker: Error response from daemon: OCI runtime create failed: invalid mount {Destination::/working_dir Type:bind Source:/var/lib/docker/volumes/1463fe5f03efda43279d6092b6735272dd97308293c7ec9adaedf57fcd9127c5/_data Options:[rbind]}: mount destination :/working_dir not absolute: unknown.
2021-06-05T14:43:43.8420517Z time="2021-06-05T14:43:43Z" level=error msg="error waiting for container: context canceled"
```
This is my task yaml:
steps:
- task: AzureStaticWebApp@0
  displayName: 'Deploy static app'
  inputs:
    app_location: '$(System.DefaultWorkingDirectory)/app/dist'
    output_location: '$(System.DefaultWorkingDirectory)/app/dist'
    skip_app_build: true
    azure_static_web_apps_api_token: '...excluded...'
```

any idea what I'm doing wrong?

**Comment by anthonychu** [comment]
@dondracek The `app_location` is relative to the working directory. It might not behave properly if you include the full path (`$(System.DefaultWorkingDirectory)`).

`output_location` is relative to the `app_location`. It's not required unless the task is building the app.

Looks like you simply want to deploy some files that have already been built. Your config should look something like this:

```yml
steps:
- task: AzureStaticWebApp@0
  displayName: 'Deploy static app'
  inputs:
    app_location: 'app/dist'
    skip_app_build: true
    azure_static_web_apps_api_token: '...excluded...'
```

If you are still stuck please create a new issue and we can help you there.

**Comment by anthonychu** [comment]
Closing this issue. Please create new issues for help or feature requests.

**Comment by ShanonJackson** [comment]
@anthonychu what about Azure Dev Ops Static Web App staging deploys? At the moment we're running two production instances using 1 as staging just to use azure dev ops.

**Comment by a2441918** [comment]
> Update on where we are today...
> 
> We're waiting for our Azure DevOps task to be published. There's still no solid ETA on this yet but hoping within the next week or so.
> 
> The portal work is ready. It'll take a week or two to make it live after the task is published.
> 
> And to set expectations, what you'll be able to do (hopefully later this month) is:
> 
> * Create a Static Web App without linking it to GitHub
> * Add an Azure Pipelines YAML to your repo with the task (we'll provide an example YAML that you can copy/paste). The task looks a lot like the GitHub Action in terms of input and capabilities. You'll need to set a pipeline variable for the deployment token (retrieved from the portal).
> * No support for PRs/environments yet. We don't have the same hooks as GitHub to act on PR open/close. We'll continue to investigate.
> 
> We'll also work on improving the experience in the coming months. But we're trying to get the basic functionality out first so most of you are unblocked ASAP. Thanks for your patience on this.

Any idea when  support for PRs/environments available??

**Comment by superelitist** [comment]
I see that this was closed after the announcement that we could create this resource through the portal. Unfortunately, I have a need to deploy a Static Web App using Azure DevOps _from the command line_. The workaround suggested by @anthonychu [here](https://github.com/Azure/static-web-apps/issues/5#issuecomment-798001564) seems to not work: I get `argument --source/-s: expected one argument`.

When can we expect support for Static Web Apps using Azure DevOps through the Azure CLI, or is there another workaround?
